test/pubspec_test.dart

```diff
@@ -639,4 +639,54 @@ features:
       });
     });
   });
+
+  group('typo warnings', () {
+    test('normal pubspec has no warnings', () {
+      final pubspec = Pubspec.fromMap({
+        'name': 'myapp',
+        'version': '1.0.0',
+        'description': 'My app.',
+        'homepage': 'https://my.homepage.com',
+        'repository': 'my-repo',
+        'issue_tracker': '',
+        'documentation': '',
+        'dependencies': {},
+        'dev_dependencies': {},
+        'dependency_overrides': {},
+        'environment': {},
+        'executables': '',
+        'publish_to': '',
+        'flutter': {}
+      }, null);
+
+      expect(pubspec.typoWarnings, equals({}));
```

Comment:
This kind of unit testing is nice for `diceCoefficient`, but for `pub` I think we generally prefer end2end integration tests..

it's also fine to have a few unit tests, but integration tests is better, because that tests what the user will experience.. Also it's what we do everywhere else in `pub`. We can't do that until we've incorporated it in the UI.

Reply:
I moved the check to the `validator` folder, and added tests in `tests/validator`, but it seems that the `lish` commands do not have end2end integration tests for the individual validators - should I add tests there nonetheless?

Reply:
now that it's a validator, I could add something like the following:

```dart
    final validator = PubspecTypoValidator(entrypoint);
    await validator.validate();

    if (validator.errors.isNotEmpty) {
      final s = validator.errors.length > 1 ? 's' : '';
      log.error('Package validation found the following error$s:');
      for (var error in validator.errors) {
        log.error("* ${error.split('\n').join('\n  ')}");
      }
      log.error('');

      flushThenExit(exit_codes.DATA);
    }
```

or even move the above into a `run()` command either for `PubspecTypoValidator` or the abstract `Validator` class, to be used in `pub get` and `pub upgrade` commands - thoughts?